### PR TITLE
feat: repo map

### PR DESCRIPTION
**Pull request by yetone**
https://github.com/yetone/avante.nvim/issues/122

### CODE REVIEW

File: lua/avante/utils/init.lua [hunk 1]
```diff
@@ -2,6 +2,8 @@ local api = vim.api
 local fn = vim.fn
 local lsp = vim.lsp
 
+local RepoMap = require("avante.utils.repo_map")
```
Comment (aarnphm) [inline lua/avante/utils/init.lua:5]: oh you can just use `M.repo_map.<func>` since L12 to L21

Comment (aarnphm) [reply lua/avante/utils/init.lua:5]: also you can add the hints to it as well 😄

File: lua/avante/utils/init.lua [hunk 2]
```diff
@@ -514,4 +516,32 @@ function M.debounce(func, delay)
   end
 end
 
+function M.get_project_root()
+  local Root = require("avante.utils.root")
+  return Root.git() or Root.cwd()
```
Comment (aarnphm) [inline lua/avante/utils/init.lua:521]:
```suggestion
	return M.root.get()
```

This should concat from lsp -> git -> cwd

File: lua/avante/api.lua [hunk 3]
```diff
@@ -115,6 +118,11 @@ M.refresh = function()
   sidebar:render()
 end
 
+M.build_project_context = function()
+  local repo_map = RepoMap.repo_map(Utils.get_project_root())
+  Path.repo_map.save(vim.api.nvim_get_current_buf(), repo_map)
```
Comment (aarnphm) [inline lua/avante/api.lua:123]:
```suggestion
  local repo_map = Utils.repo_map.repo_map(Utils.root.get())
  Path.repo_map.save(vim.api.nvim_get_current_buf(), repo_map)
```

File: lua/avante/init.lua [hunk 4]
```diff
@@ -43,6 +43,11 @@ H.commands = function()
     function(opts) require("avante.api").edit(vim.trim(opts.args)) end,
     { desc = "avante: edit selected block", nargs = "*" }
   )
+  cmd(
+    "BuildProjectContext",
+    function() require("avante.api").build_project_context() end,
+    { desc = "avante: build project context", nargs = "*" }
+  )
```
Comment (aarnphm) [inline lua/avante/init.lua:50]: I think we can build the project context asynchronously?

File: lua/avante/utils/repo_map.lua [hunk 5]
```diff
@@ -0,0 +1,526 @@
+local parsers = require("nvim-treesitter.parsers")
+
+---@class avante.utils.repo_map
+local RepoMap = {}
+
+local dependencies_queries = {
+  lua = [[
```
Comment (aarnphm) [inline lua/avante/utils/repo_map.lua:7]: hmm, are we down to maintain query for all supported language?

Then I suggest we should use query file instead?

**Comment by yetone** [comment]
This PR is still in rapid iteration and is not yet ready for review.

File: lua/avante/utils/repo_map.lua (hunk 5 above)
Comment (yetone) [reply lua/avante/utils/repo_map.lua:7]: @aarnphm My brain is limited, I need to implement quickly, and think about project structure issues after implementation is complete.

**Comment by jmmarotta** [comment]
You could probably use the [aider queries](https://github.com/paul-gauthier/aider/tree/main/aider/queries) for treesitter.

**Comment by aarnphm** [comment]
we are iterating a lot atm, and will consider refactor the query later.

**Comment by yetone** [comment]
> You could probably use the [aider queries](https://github.com/paul-gauthier/aider/tree/main/aider/queries?rgh-link-date=2024-09-05T19%3A57%3A44Z) for treesitter.您可能可以使用树木管理员的[辅助查询](https://github.com/paul-gauthier/aider/tree/main/aider/queries)。

aider's queries return all class/function definitions in the file, but I only need the externally exposed classes/functions (similar to a C header file), so their code doesn't suit my requirements.

**Comment by yetone** [comment]
Here's an update for everyone: Actually, this PR was completed a long time ago. The functionality tests are all normal and the results are very good, but it consumes a large number of tokens (to test this PR, my Anthropic quota has already been exhausted 😢). We are currently considering how to optimize this aspect.

**Comment by SchnozzleCat** [comment]
> Here's an update for everyone: Actually, this PR was completed a long time ago. The functionality tests are all normal and the results are very good, but it consumes a large number of tokens (to test this PR, my Anthropic quota has already been exhausted 😢). We are currently considering how to optimize this aspect.

Is the consumed token amount similar to what is consumed when using something like Aider (and adding all files in a project)?

**Comment by jmmarotta** [comment]
> > Here's an update for everyone: Actually, this PR was completed a long time ago. The functionality tests are all normal and the results are very good, but it consumes a large number of tokens (to test this PR, my Anthropic quota has already been exhausted 😢). We are currently considering how to optimize this aspect.
> 
> Is the consumed token amount similar to what is consumed when using something like Aider (and adding all files in a project)?

The core issue seems to be that the current implementation is using the entire repomap instead of focusing on the most relevant tags. A potential solution would involve finding a way to rank the tags produced by tree-sitter.

In Python ecosystems, libraries like NetworkX (used in aider) can represent these tags as nodes in a graph and then apply algorithms like PageRank to rank them effectively.

**Comment by aarnphm** [comment]
afaik nvim has binding to python, but it is kinda slow.

Maybe we can do reranking through colbert from rust/c++ or sth.

**Comment by jmmarotta** [comment]
Pagerank might work better than colbert as it's well-suited for treesitter's tag structure, but using rust/c++ is a great idea.

[petgraph](https://github.com/petgraph/petgraph) might work well

**Comment by aarnphm** [comment]
> [petgraph](https://github.com/petgraph/petgraph) might work well

Thanks for the knowledge sharing, will take a look into this

**Comment by yetone** [comment]
> > Is the consumed token amount similar to what is consumed when using something like Aider (and adding all files in a project)?
> 
> 
> 
> The core issue seems to be that the current implementation is using the entire repomap instead of focusing on the most relevant tags. A potential solution would involve finding a way to rank the tags produced by tree-sitter.
> 
> 
> 
> In Python ecosystems, libraries like NetworkX (used in aider) can represent these tags as nodes in a graph and then apply algorithms like PageRank to rank them effectively.

In my local tests, aider's repo map tends to be inaccurate, bringing in unrelated files while ignoring related ones.

**Comment by jmmarotta** [comment]
> In my local tests, aider's repo map tends to be inaccurate, bringing in unrelated files while ignoring related ones.

PageRank likely overweights common utilities. For a more localized context BFS might be the best.

**Pull request merged**